picard/cluster.py

```diff
@@ -166,6 +169,17 @@ def remove_file(self, file, new_album=True):
 
     def update(self, signal=True):
         self.metadata['~totalalbumtracks'] = self.metadata['totaltracks'] = len(self.files)
+        metadata_counters = {}
+        tags_to_update = ['catalognumber', 'barcode', 'asin', 'genre']
+        for file in self.files:
+            for tag in tags_to_update:
+                if tag in file.metadata:
+                    value = file.metadata[tag]
+                    if tag not in metadata_counters:
+                        metadata_counters[tag] = Counter()
+                    metadata_counters[tag][value] += 1
+        for tag in metadata_counters:
+            self.metadata[tag] = metadata_counters[tag].most_common(1)[0][0]
```

Comment:
What is the purpose of this additional code (which is only for these tags)?

And why has genre been added to these tags?

Reply:
> What is the purpose of this additional code (which is only for these tags)?

This is in order to add those tags to the cluster so they can be used in the lookup.

> And why has genre been added to these tags?

Because I thought it would make sense for a cluster to also have the most common genre in its files, there may also be other tags which makes sense to include (like `releasecountry` or `comment` for release disambiguation maybe?) and all these can be quickly added/removed to the lookup (which could also lead to PICARD-2793)

Reply:
Additional code - I haven't read the existing code in detail, but I surely code to include these tags in the cluster should be co-located with the code that puts the existing tags into the cluster.

Genres should not IMO be added because they are highly subjective and IMO the MB implementation is far too free-form i.e. lacking in hierarchy and unstructured.

Reply:
Hmm, I had searched for that a bit after trying to just add `catno=self.metadata['catalognumber'],` to the `find_releases` call in `Cluster.lookup_metadata()` (which didn't work) but I couldn't find anything.

Now I tried (on `master`) printing `self.metadata` right before the `find_releases` call and the log just reads
```
store: {'album': ['Do the Dive [Limited Edition]'], 'albumartist': ['Call of Artemis'], 'totaltracks': ['4'], '~totalalbumtracks': ['4']}
deleted: set()
images: ["TagCoverArtImage from '/I/Raccolte/Musica/D4DJ Collection/Call of Artemis - Do the Dive [Limited Edition] (2022) [FLAC+CUE+LOG] {BRMM-10544}/01. Do the Dive.flac' of type front"]
length: 1066065
```
(all 4 files have `catalognumber` set)
So it doesn't seem to get populated with the catalog number.
Unless I didn't understand something and those tag shouldn't be in `self.metadata`

Now that I think about it tho, in the UI I can see the cluster files' metadata, I'm gonna search if I can find where it gets them from.

Reply:
I think generally this tag merging for clusters makes sense, and we should do something like this for some relevant tags. It makes the clusters far more useful if we add such lookups. Whether this includes genres or not is a different question, and currently I don't have any strong opinion on this either way.

I'm a bit concerned about performance here if this runs for every individual file added. But this would need to be measured. Optimization would be to ensure this runs only after all files have been added when clustering multiple files at once (i.e. usually when creating the cluster).

Reply:
> I'm a bit concerned about performance here if this runs for every individual file added. But this would need to be measured. Optimization would be to ensure this runs only after all files have been added when clustering multiple files at once

I was also concerned about that, but from what I saw, `self.update()` gets called only once after all files have been added (after `for file in files` loop) in `cluster.add_files()`, so it should run only one time for each cluster (as you said it needs to be tested tho).

> (i.e. usually when creating the cluster).

Actually my first approach was to do this, by passing the metadata objects of the various files to `FileCluster`[1] and have a `FileCluster.metadata` property returning a metadata object with the most common tags, but that would require quite a bit of changes, also in the tests (as it would change the arguments of `FileCluster.add()` and  `Cluster`'s constructor).

I still have these changes stashed, I'll push them to a separate branch, if you want to see them (they should work, but the tests didn't all pass and there are still some problems).

[1] like this
```diff
# in cluster.py, Cluster.cluster()
    @staticmethod
    def cluster(files):
        # ...
        for file in files:
            # ...
            token = tokenize(album)
            if token:
                # Basically pass `metadata` here
-               cluster_list[token].add(album, artist or various_artists, file)
+               cluster_list[token].add(metadata, file)
```